### PR TITLE
Authentication

### LINKED ISSUE #41: Authentication
#### As an application hoster I want to add user authentication to my application.
- [ ] Users can sign up.
- [ ] Users can log in after they've signed up.
- [ ] Users can log out after logging in.
- [ ] Users can choose to be remembered after logging in.
- [ ] Multiple browsers can concurrently hold permanent cookies.
- [ ] Logged in users should not be able to register, nor log in again.

Discussion topics:
- [ ] If we wanted to add more advanced features such as email activation and
  "forgot my password", should we continue to "roll our own" authentication?

Reference material:
- https://www.railstutorial.org/book/log_in_log_out

Further reading:
- https://github.com/plataformatec/devise

Dependencies:
- Delete Image

### PR DESCRIPTION
**Pull request by HaominZhang**
Closes #41

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -34,7 +34,7 @@ gem 'bootstrap', '~> 4.0.0.alpha3'
 gem 'rails-assets-tether', '~> 1.1.0'
 
 # Use ActiveModel has_secure_password
-# gem 'bcrypt', '~> 3.1.7'
+gem 'bcrypt', '~> 3.1.7'
```
Comment (bboe) [inline Gemfile]: :+1: way to find it in the commented out section.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -2,4 +2,5 @@ class ApplicationController < ActionController::Base
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+  include SessionsHelper
```
Comment (bboe) [inline app/controllers/application_controller.rb]: Below is a copy-paste the comment @rmacklin made to Boying about this same thing:

> SessionHelper is intended to be used as the view helper for the SessionController's views, and it's actually an antipattern to include that helper here in ApplicationController. I realize that this was actually suggested in the Rails Tutorial book, but I wanted to mention it.

> A way you could work around this is to extract the methods that you need access to outside of the SessionController views into another module, and include that module where you need it as well as in SessionHelper. A benefit of that is that then you'll only have access to the methods you actually want available exactly where you need them, rather than having access to all the methods in SessionHelper in ever controller that inherits from ApplicationController.

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+class SessionsController < ApplicationController
+  def new
+    redirect_to root_path if logged_in?
+  end
+
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
```
Comment (bboe) [inline app/controllers/sessions_controller.rb]: Is an SQL query executed when email isn't provided?

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    if user
+      if user.authenticate(params[:session][:password])
+        log_in user
+        params[:session][:remember_me] == '1' ? remember(user) : forget(user)
+        redirect_to root_path
+      else
+        redirect_to :login, notice: 'Invalid email/password combination, please try again.'
```
Comment (bboe) [inline app/controllers/sessions_controller.rb]: This uses a different behavior than create Image when the form fields are "invalid". Can we keep the approach consistent?

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  def new
+    redirect_to root_path if logged_in?
+  end
+
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
+    if user
+      if user.authenticate(params[:session][:password])
+        log_in user
+        params[:session][:remember_me] == '1' ? remember(user) : forget(user)
```
Comment (bboe) [inline app/controllers/sessions_controller.rb]: What are the implications of this `forget(user)` if a user has more than one computer they regularly use?

File: app/controllers/sessions_controller.rb (hunk 3 above)
Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: I am not pretty sure but I guess the answer is yes. Should I try to avoid the SQL executed if the email isn't provided?

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
+class UsersController < ApplicationController
+  def new
+    if !logged_in?
```
Comment (bboe) [inline app/controllers/users_controller.rb]: Awesome!

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        redirect_to root_path
+      else
+        redirect_to :login, notice: 'Invalid email/password combination, please try again.'
+      end
+    else user.nil?
+      redirect_to :login, notice: "The email you've entered doesn't match any account."
```
Comment (bboe) [inline app/controllers/sessions_controller.rb]: What are the security implications of having this separate notice?

File: app/helpers/application_helper.rb [hunk 8]
```diff
@@ -19,4 +23,14 @@ def navigation_item(controller_name)
       link_to controller_name.titlecase, send("#{controller_name}_path"), class: "nav-link#{ ' active' if params[:controller] == controller_name}"
     end
   end
+
+  def print_user_name(user_name)
```
Comment (bboe) [inline app/helpers/application_helper.rb]: Is `print` the right verb to use?

Comment (HaominZhang) [inline app/helpers/application_helper.rb]: You're right. Maybe `display` is better?

File: app/helpers/application_helper.rb [hunk 9]
```diff
@@ -19,4 +23,14 @@ def navigation_item(controller_name)
...
+    content_tag(:li, "#{user_name}", class: 'nav-item__name nav-link active')
+  end
+
+  def print_logout_button
+    content_tag(:li, class: 'nav-item__button') do
+      link_to 'Log out', send("logout_path"), method: :delete, class: 'nav-link'
```
Comment (bboe) [inline app/helpers/application_helper.rb]: Is `send` required here?

File: app/helpers/sessions_helper.rb [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    cookies.delete(:user_id)
+    cookies.delete(:remember_token)
+  end
+
+  def current_user
+    if (user_id = session[:user_id])
+      @current_user ||= User.find_by(id: user_id)
+    elsif (user_id = cookies.signed[:user_id])
+      user = User.find_by(id: user_id)
+      if user && user.authenticated?(cookies[:remember_token])
```
Comment (bboe) [inline app/helpers/sessions_helper.rb]: What should we do if `user` doesn't exist?

File: app/helpers/application_helper.rb (hunk 9 above)
Comment (HaominZhang) [inline app/helpers/application_helper.rb]: Yeah, without the send, the path will not work

File: app/controllers/sessions_controller.rb (hunk 4 above)
Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: Nice catch I will fix this

File: app/controllers/sessions_controller.rb (hunk 7 above)
Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: As we discussed yesterday, there is some information exposed here. I will fix this.

File: app/helpers/sessions_helper.rb (hunk 10 above)
Comment (HaominZhang) [inline app/helpers/sessions_helper.rb]: Good point, I will fix it

File: app/controllers/sessions_controller.rb [hunk 11]
```diff
@@ -0,0 +1,23 @@
+class SessionsController < ApplicationController
+  include SessionsHelper
+  def new
+    redirect_to root_path if logged_in?
+  end
+
+  def create
+    user =  params[:session][:email].present? && User.find_by(email: params[:session][:email].downcase)
```
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: What will happen if a user is already logged in and tries to re-authenticate?

Comment (d-ray) [inline app/controllers/sessions_controller.rb]: :space_invader: (ask @bboe)

File: app/controllers/sessions_controller.rb [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    if user && user.authenticate(params[:session][:password])
+      log_in user
```
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: Will the user have any indication that they've successfully logged in?

File: app/controllers/sessions_controller.rb [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+      remember(user) if params[:session][:remember_me] == '1'
+      redirect_to root_path
+    else
+      flash[:notice] = 'Invalid email/password combination, please try again.'
```
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: Is notice the appropriate type of flash message for this?  Are you distinguishing between different types of messages (e.g. notice vs. error)?

File: app/controllers/sessions_controller.rb [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+      render :new
```
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: This will rerender a blank form.  Would it be more user friendly to pre-populate some/all of the info they entered?

File: app/controllers/sessions_controller.rb [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+    end
+  end
+
+  def destroy
+    log_out if logged_in?
+    redirect_to login_path
```
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: It's common that logging out directs a user to the login page, but that may not be the best experience in this scenario.  Another option would be to keep the user on the same page.  How do you think those two options compare for this application?

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -0,0 +1,27 @@
+class UsersController < ApplicationController
+  include LoginHelper
+  def new
+    if !logged_in?
```
Comment (d-ray) [inline app/controllers/users_controller.rb]: I often find that checks with out negation are easier to understand.  Perhaps it would be clearer to switch the order of the `if`/`else` branches?

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+      @user = User.new
+    else
+      redirect_to root_path
+    end
+  end
+
+  def create
+    @user = User.new(user_params)
```
Comment (d-ray) [inline app/controllers/users_controller.rb]: Should you also check whether the user is logged in here (as in `new` above)?

File: app/helpers/application_helper.rb [hunk 18]
```diff
@@ -10,7 +10,13 @@ def navigation_bar(brand:, items:)
 
   def navigation_items(*controllers)
     content_tag(:ul, class: 'nav navbar-nav') do
-      controllers.map { |controller| navigation_item(controller) }.join.html_safe
+      html = controllers.map { |controller| navigation_item(controller) }.join.html_safe
+      if logged_in?
+        html << display_logout_button << display_user_name(@current_user.name)
```
Comment (d-ray) [inline app/helpers/application_helper.rb]: Do you find that using `<<` multiple times on a single line promotes optimal readability?

File: app/helpers/application_helper.rb [hunk 19]
```diff
@@ -10,7 +10,13 @@ def navigation_bar(brand:, items:)
 
   def navigation_items(*controllers)
     content_tag(:ul, class: 'nav navbar-nav') do
-      controllers.map { |controller| navigation_item(controller) }.join.html_safe
+      html = controllers.map { |controller| navigation_item(controller) }.join.html_safe
```
Comment (d-ray) [inline app/helpers/application_helper.rb]: Is it necessary to call `html_safe` here?

File: app/helpers/application_helper.rb [hunk 20]
```diff
@@ -10,7 +10,13 @@ def navigation_bar(brand:, items:)
...
+      if logged_in?
+        html << display_logout_button << display_user_name(@current_user.name)
+      else
+        html << display_login_button
```
Comment (d-ray) [inline app/helpers/application_helper.rb]: Would it be worthwhile to add the register button here as well?

Comment (HaominZhang) [inline app/helpers/application_helper.rb]: I can definitely add the register button but which design do you prefer? I checked both FB and Instagram, and for FB, if I access some page without logging in, I can see a register button on the navbar but for Instagram, there is only a log in button. So which one seems more proper?

File: app/helpers/application_helper.rb (hunk 18 above)
Comment (HaominZhang) [inline app/helpers/application_helper.rb]: So maybe `html << display_logout_button`and then`html << display_user_name` is better?

File: app/controllers/sessions_controller.rb (hunk 13 above)
Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: Actually I didn't distinguish different types of messages but definitely I should do it. I will fix it.

File: app/controllers/users_controller.rb (hunk 16 above)
Comment (HaominZhang) [inline app/controllers/users_controller.rb]: Good point, I will fix this.

File: app/controllers/sessions_controller.rb (hunk 14 above)
Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: This is a little weird to me because in here https://github.com/AppFolioOnboarding/Haomin_ImageUploader/blob/master/app/controllers/pictures_controller.rb#L37, I also use render :new, together with status code, but the new rendered form is pre-populated with info entered. However, even if I add status code after this `render :new`, it still doesn't work. Do you know the reason for that?

File: app/controllers/sessions_controller.rb (hunk 15 above)
Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: Nice catch, I guess redirecting to the same page would be a better option. I will fix this.

File: app/controllers/users_controller.rb (hunk 17 above)
Comment (HaominZhang) [inline app/controllers/users_controller.rb]: I discussed this with @rmacklin and the case I tested is that I opened two tabs both on login page and log in with two accounts one by one and both of them can successfully log in. But if I refresh the page of first logged-in account, it will become the second logged-in account, which seems to acceptable. How do you think about it?

File: app/helpers/application_helper.rb (hunk 19 above)
Comment (HaominZhang) [inline app/helpers/application_helper.rb]: I think you are right, since I called html.safe at the very bottom, I don't have to call it here. I will fix this.

File: app/controllers/sessions_controller.rb (hunk 14 above)
Comment (rmacklin) [inline app/controllers/sessions_controller.rb]: @HaominZhang I'll give you a hint: Check the ERB templates for the picture form and compare it to the log in form.

File: app/helpers/application_helper.rb (hunk 20 above)
Comment (rmacklin) [inline app/helpers/application_helper.rb]: I think either is fine - it's up to you. Your login page includes a link to sign up, which satisfies the acceptance criteria for being able to register a new account.

Comment (d-ray) [inline app/helpers/application_helper.rb]: :+1: to Richard's comment.  You don't have to add it; it was just a suggestion/something to think about.  It's great that you're looking around to see what others are doing though.

File: app/helpers/application_helper.rb (hunk 18 above)
Comment (d-ray) [inline app/helpers/application_helper.rb]: Personally, I find that to be clearer.  But ultimately it's your choice.

File: app/controllers/users_controller.rb (hunk 17 above)
Comment (d-ray) [inline app/controllers/users_controller.rb]: That's probably acceptable behavior.  I would expect it to prevent the second user from logging in (just as you would be prevented from going to the log in page if a user is already logged in).

Comment (bboe) [inline app/controllers/users_controller.rb]: @HaominZhang I think it should be programmed to not allow a second login. The two-tab situation you described shouldn't be permitted.

Comment (rmacklin) [inline app/controllers/users_controller.rb]: Really? I think the two tab scenario is an edge case and the behavior of having the "last login win" is fine. I just tried this with Gmail and that's the behavior they have (well, they also pop up an alert in the first tab telling you you've been logged out, but that's definitely overkill). Given that this is just a learning exercise to get _some_ auth in place for the favoriting story, I'd lean towards letting him leave it as is unless you feel really strongly.

Comment (bboe) [inline app/controllers/users_controller.rb]: While this is an edge case, I think we can both agree the functionality should either be completely supported or completed unsupported. Leaving it in an "undefined" behavior state could, if this were a real project, result in hard-to-track down bugs. Requiring an explicit logout before subsequent login seems simpler.

Alternatively, if leaving the login POST open is desired, then I think access to the login NEW action should remain for consistency (though the link to it should not).

Comment (bboe) [inline app/controllers/users_controller.rb]: A non-trivial example of undefined behavior would be if a certain action set a custom cookie for a logged in user. It would be expected that such a cookie would be deleted along with the session when the user logged out.

However, if login-when-login can occur, but isn't a first-class action, then that cookie meant for the previous user may not get cleaned up when logging in for the second time.

I guess the two options are (I'm fine with either) to implicitly log out a user if there is one (assuming all the clean-up occurs there) before logging in the new user, or to (more simply IMO) not permit the second login.

File: app/controllers/sessions_controller.rb (hunk 11 above)
Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: I confirmed with @bboe and I will fix this.

File: app/controllers/sessions_controller.rb (hunk 12 above)
Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: Currently just displaying the user name on navbar. I think it is enough for right now based on the story issues. How do you think?

File: app/controllers/sessions_controller.rb (hunk 14 above)
Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: @rmacklin Oh I found the difference. In log in form, the model used for simple form is `:session`, but `:session` is not either an ActiveRecord nor an ActiveModel, which means that there is no validator there. So apparently there will be no validation in this form. Is that right? But how can I resolve this issue if I want to keep the info that user entered before?

File: app/helpers/application_helper.rb (hunk 18 above)
Comment (HaominZhang) [inline app/helpers/application_helper.rb]: :+1:

File: app/controllers/sessions_controller.rb (hunk 14 above)
Comment (rmacklin) [inline app/controllers/sessions_controller.rb]: You're on the right track. `simple_form_for` can take an object such as a model, or a symbol. If the latter, the symbol is just used as the namespace for the form fields (So, `simple_form_for :session` with `f.input :email` will create a form input with the name `session[:email]`, which is what ultimately gets you `params[:session][:email]` in your controller). When you pass it an object (e.g. an ActiveRecord model or a form object), it'll use that to get the namespace as well as to fill in values for the generated `input`s using whatever is currently in those attributes of your object.

> which means that there is no validator there. So apparently there will be no validation in this form. Is that right?

Validations don't factor into the above behavior. `simple_form` will use the `errors` object to add its inline error messages (typically coming from validations), but that independent of autofilling the `input` values.

> But how can I resolve this issue if I want to keep the info that user entered before?

What ideas do you have?

Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: I guess we can use JavaScript to fill the form when we render this page. However, I also think there is simpler way to do it..

Comment (rmacklin) [inline app/controllers/sessions_controller.rb]: > However, I also think there is simpler way to do it

I agree :wink:. Have you made any other forms that you could use as an example?

Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: I found out a solution like this: 
`<%= u.input :email, input_html: { value: params[:session].present? ? params[:session][:email] : nil }%>`
  `<%= u.input :password, input_html: { value: params[:session].present? ? params[:session][:password]: nil } %>`
I also used `input_html` in update tag list page, where the tags are auto-filled. Is this the solution that you would like to point out?

Comment (rmacklin) [inline app/controllers/sessions_controller.rb]: That works! A few comments:

Expressions like: `params[:session].present? ? params[:session][:email] : nil` can be reduced to an equivalent but more concise form via: `params[:session].presence`

While the above works, did you look at how you solved this problem for the share image form?

Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: In share image form, if anything goes wrong in the modal, I will respond a json object in which I render an html as string and pass a `@share_form`.
I tried to do the similar things but I failed..

Comment (rmacklin) [inline app/controllers/sessions_controller.rb]: > In share image form, if anything goes wrong in the modal, I will respond a json object

You're starting to look in the wrong places. All I'm asking is about the form. What did you pass to `simple_form_for` there to make it remember your input values?

Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: I passed a ShareForm instance that is created when I try to render the partial within modal body

Comment (rmacklin) [inline app/controllers/sessions_controller.rb]: Right. So you could similarly create a new form object and pass that

Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: So are you saying I should create a class for session just as I created a class form ShareForm? And then I can do the similar way passing a new form object to the log in form?

Comment (rmacklin) [inline app/controllers/sessions_controller.rb]: Yeah, I'm saying you could create a `LoginForm` class similar to how you created the `ShareForm` class.

Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: Oh I got you. But just want to confirm, compared with solution I found above using `input_html`, the solution you suggested seems a little more complicated because I have to create another class and also pass a new instance to the partial rather than just adding two lines in my html. So is the way you suggested more typical to solve this kind of issue or actually I can take either one as the solution in this case?

Comment (rmacklin) [inline app/controllers/sessions_controller.rb]: The way I suggest is how we'd typically do it - it's cleaner than reaching into `params` from the view. And I think you'll find that the overhead of the extra class is _very_ minimal.

File: app/controllers/sessions_controller.rb (hunk 12 above)
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: It might be nice to have a flash message confirming a successful login since the app looks almost the same whether you're logged in or not.  It's up to you though; as you noted, it's not part of the acceptance tests for the story.  I mention it mainly as something to think about; ideally the acceptance tests fully enumerate what needs to be done for a story to be complete, but in practice it's not uncommon that edge cases or small usability points like this aren't detailed in the relevant story.  That being the case, it's good to be in the habit of asking these kinds of questions as you transition out of Ropes and on to a regular development team.

Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: Yeah you are right, it would be nicer to have a flash message and also I should make it consistent with logged in or not. I will fix this.

File: app/controllers/sessions_controller.rb (hunk 13 above)
Comment (HaominZhang) [inline app/controllers/sessions_controller.rb]: I will create another branch to distinguish the type of messages after this branch

File: app/controllers/concerns/login_concern.rb [hunk 21]
```diff
@@ -0,0 +1,27 @@
+module LoginConcern
+  def log_in(user)
+    session[:user_id] = user.id
+  end
+
+  def current_user
+    if (user_id = session[:user_id])
+      @current_user ||= User.find_by(id: user_id)
+    elsif (user_id = cookies.signed[:user_id])
+      user = user_id.present? && User.find_by(id: user_id)
```
Comment (d-ray) [inline app/controllers/concerns/login_concern.rb]: Do you think variable assignment involving `&&` promotes optimal readability?

File: app/controllers/concerns/login_concern.rb [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+      if user
+        log_in user
+        @current_user = user
+      else
+        forget user
```
Comment (d-ray) [inline app/controllers/concerns/login_concern.rb]: What's the purpose of passing `user` as a param to `forget`?  On a related note, is this branch of the conditional adequately tested?

File: app/controllers/sessions_controller.rb [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+
+  private
+
+  def remember(user)
+    cookies.permanent.signed[:user_id] = user.id
+  end
+
+  def log_out
+    session.delete(:user_id)
+    forget
```
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: Does this method name adequately communicate what the method does (particularly when it's used outside of the context of the `LoginConcern`)?

File: app/models/user.rb [hunk 24]
```diff
@@ -0,0 +1,25 @@
+class User < ActiveRecord::Base
+  MIN_PASSWORD_LENGTH = 6
+  MAX_NAME_LENGTH     = 50
+  MAX_EMAIL_LENGTH    = 255
+
+  before_save { self.email = email.downcase }
+  validates :name, presence: true, length: { maximum: MAX_NAME_LENGTH }
+  VALID_EMAIL_PATTERN = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
```
Comment (d-ray) [inline app/models/user.rb]: Should this go at the top of the file with the other constants?

File: app/controllers/sessions_controller.rb [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+    else
+      user = params[:login_form][:login_email].present? && User.find_by(email: params[:login_form][:login_email].downcase)
+      if user && user.authenticate(params[:login_form][:login_password])
+        log_in user
+        remember(user) if params[:login_form][:login_remember_me] == '1'
+        redirect_to root_path, notice: 'Successfully Logged in! Enjoy!'
+      else
+        flash[:notice] = 'Invalid email/password combination, please try again.'
+        @session_object = LoginForm.new(params[:login_form])
+        render :new, object: @session_object
```
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: Do you need to pass `@session_object` as `object` of is the instance variable sufficient?

File: app/controllers/sessions_controller.rb [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+
+  def new
+    redirect_to root_path if logged_in?
+  end
+
+  def create
+    if logged_in?
+      redirect_to root_path, notice: 'You already logged in, you can not log in with another account before logging out.'
+    else
+      user = params[:login_form][:login_email].present? && User.find_by(email: params[:login_form][:login_email].downcase)
```
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: Do you think variable assignment involving && promotes optimal readability?

File: test/controllers/sessions_controller_test.rb [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+    params = { login_email: 'email@example.com', login_password: 'test_password', login_remember_me: '0' }
+    post :create, login_form: params
+    assert_predicate cookies.permanent.signed[:user_id], :nil?
+  end
+
+  test 'cannot access login page after login' do
+    session[:user_id] = @user.id
+    get :new
+    assert_redirected_to root_path
+  end
```
Comment (d-ray) [inline test/controllers/sessions_controller_test.rb]: Is this code path tested? https://github.com/AppFolioOnboarding/Haomin_ImageUploader/pull/50/files#diff-d5241d488259f32ecbe2f636133e5ddaR15

File: test/controllers/users_controller_test.rb [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+  test 'cannot access sign up page after login' do
+    user = User.create!(name: 'test_name',
+                        email: 'email@example.com',
+                        password: 'test_password',
+                        password_confirmation: 'test_password')
+
+    session[:user_id] = user.id
+    get :new
+    assert_redirected_to root_path
+  end
```
Comment (d-ray) [inline test/controllers/users_controller_test.rb]: Is this code path tested?  https://github.com/AppFolioOnboarding/Haomin_ImageUploader/pull/50/files#diff-4e05ad0d64e6100656b63ad1e78f32c5R14

**Comment by bboe** [comment]
Closing this to remove it from my list of mentioned pull requests.
